### PR TITLE
Do one round of DestProp at mir-opt-level=2

### PR DESCRIPTION
**Pull request by saethlin**
Based on some brief experimentation with DestinationPropagation, it seems like sometimes we run a lot of rounds, where the first round does a _lot_ more than any of the later rounds. So it seems like doing just the first round is probably a much better compile time tradeoff than doing all of them.

### CODE REVIEW

**Comment by saethlin** [comment]
@JakobDegen Based on the perf run above, are you interested in this?

File: compiler/rustc_mir_transform/src/dest_prop.rs [hunk 1]
```diff
@@ -157,7 +157,7 @@ impl<'tcx> MirPass<'tcx> for DestinationPropagation {
         //  2. Despite being an overall perf improvement, this still causes a 30% regression in
         //     keccak. We can temporarily fix this by bounding function size, but in the long term
         //     we should fix this by being smarter about invalidating analysis results.
-        sess.mir_opt_level() >= 3
+        sess.mir_opt_level() >= 2
```
Comment (MinerSebas) [inline compiler/rustc_mir_transform/src/dest_prop.rs:160]: The first line of the comment above isn't adjusted to mention the new behavior.

**Comment by saethlin** [comment]
FYI I have no idea why the codegen tests are failing in CI, they pass locally and I can't reproduce the CI failure. I've tried flipping quite a few options in my config and none seem to produce the test failure.

**Comment by saethlin** [comment]
r? rust-lang/compiler

**Comment by oli-obk** [comment]
oh wow, that's an amazing find. typeck and borrowck actually get faster because this causes rustc to improve, offsetting regressions which do work in optimized_mir.

@bors r+ rollup=never

**Comment by tmiasko** [comment]
We should do a crater run before enabling the pass by default. There are also a known issues that should be addressed first: #105428, #105466, and stack space regression described in the comments.

**Comment by oli-obk** [comment]
@craterbot check

**Comment by JakobDegen** [comment]
@saethlin https://github.com/rust-lang/rust/pull/96451#discussion_r919788899 is the reproducer. That might not work anymore with this version, but let me know if it doesn't and I can make one that will

**Comment by saethlin** [comment]
I responded on that comment thread

**Comment by saethlin** [comment]
I'm not an expert on crater, but the ICE in https://github.com/rust-lang/rust/issues/105428 needs `-Zmir-opt-level=2`, which in this case is a release build. So I don't think `craterbot check` can tickle the ICEing code path, and do we even have a way to do a `--release` build in crater?

**Comment by JakobDegen** [comment]
Oh, yeah craterbot check won't execute this codepath at all I think...

**Comment by saethlin** [comment]
Here's a demonstration of the stack size issue, unfortunately I can't pin to a single nightly so this continues to demonstrate the problem: https://godbolt.org/z/fhjc1qajM

**Comment by oli-obk** [comment]
@craterbot build

**Comment by oli-obk** [comment]
@craterbot run mode=build-only

**Comment by saethlin** [comment]
The crater run found nothing, as expected. We would need a way to ensure that we have `-Zmir-opt-level=2 -Zvalidate-mir` on in crater. I can push a hacky commit that makes those the defaults?

**Comment by oli-obk** [comment]
I flipped crater to build mode, so that should have run the changes from your PR I think?

**Comment by saethlin** [comment]
I'm pretty sure crater's build mode is a debug build, which is `-Zmir-opt-level=1`. Even with this PR, DestProp is still a `-Zmir-opt-level=2`-only pass. (and we would need `-Zvalidate-mir`)

It looks like craterbot supports setting rustflags: https://github.com/rust-lang/crater/blob/master/docs/bot-usage.md#experiment-requirements

**Comment by JakobDegen** [comment]
If we do another run, let's wait for the changes I have pending to land first, specifically the fix for stack size and dealing with moves

**Comment by saethlin** [comment]
Waiting on DestProp improvement
@rustbot label +S-blocked

**Comment by saethlin** [comment]
Since https://github.com/rust-lang/rust/pull/107449, it seems highly unlikely that any applying any kind of cleverness like this will let DestinationPropagation be a compile time improvement.

So I think the status quo is fine and there's nothing to do here: DestinationPropagation can stay opt-in, for things like cranelift.

**Pull request closed without merging**